src/pybind/mgr/volumes/fs/operations/versions/subvolume_v1.py

```diff
@@ -30,18 +22,19 @@ class SubvolumeV1(SubvolumeBase, SubvolumeTemplate):
     def version():
         return SubvolumeV1.VERSION
 
-    @property
-    def path(self):
-        try:
-            # no need to stat the path -- open() does that
-            return self.metadata_mgr.get_global_option(MetadataManager.GLOBAL_META_KEY_PATH).encode('utf-8')
-        except MetadataMgrException as me:
-            raise VolumeException(-errno.EINVAL, "error fetching subvolume metadata")
-
     @property
     def features(self):
         return [SubvolumeFeatures.FEATURE_SNAPSHOT_CLONE.value, SubvolumeFeatures.FEATURE_SNAPSHOT_AUTOPROTECT.value]
 
+    def snapshot_base_path(self):
+        return os.path.join(self.path, self.vol_spec.snapshot_dir_prefix.encode('utf-8'))
+
+    def snapshot_path(self, snapname):
+        return os.path.join(self.snapshot_base_path(), snapname.encode('utf-8'))
+
+    def snapshot_data_path(self, snapname):
```

Comment:
can you explain a bit more about snapshot_data_path() in comparison to snapshot_path()?

Reply:
NOTE (digression): I should add (in the doc commit) comments on what is a v1/2 subvolume and the explanation that suits various snapshot paths as per this discussion

Initial comments on the various snapshot paths are provided [here](https://github.com/ceph/ceph/pull/35647/files#diff-a46bf7683125b6992877591c70815e01R115-R128) and I am going to build on these.

snapshot_base_path(self): Base path for all snapshots, for a subvolume
snapshot_path(snapname): Path to a specific snapshot named 'snapname', for a subvolume
snapshot_data_path(snapname): Path to user data directory within a subvolume snapshot named 'snapname'

snapshot_path is a path to a CephFS snapshot named snapname under the subvolume, and snapshot_data_path is the path within a CephFS snapshot under a subvolume, that contains user data.

As you know, in v1 both paths are the same. In v2, they are different as the data_path is the incarnation directory within the CephFS snapshot.

:+1: If a better name suits snapshot_data_path, will gladly change it.

Reply:
Thanks for explaining -- I'll relook at the snapshot_data_path() usage.